x-pack/plugins/gis/public/shared/layers/sources/es_geohashgrid_source.js

```diff
@@ -87,7 +96,61 @@ export class ESGeohashGridSource extends ASource {
     );
   }
 
+
+  async getGeoJsonWithMeta({ layerId, layerName }, searchFilters) {
+
+    let targetPrecision = ZOOM_TO_PRECISION[Math.round(searchFilters.zoom)];
+    targetPrecision += 0;//should have refinement param, similar to heatmap style
+    const featureCollection = await this.getGeoJsonPointsWithTotalCount({
+      precision: targetPrecision,
+      extent: searchFilters.buffer,
+      timeFilters: searchFilters.timeFilters,
+      layerId: layerId,
+      layerName: layerName
+    });
+
+    if (this._descriptor.requestType === REQUEST_TYPE.AS_CENTROID_POINT) {
+      //can use as-is
+    } else if (this._descriptor.requestType === REQUEST_TYPE.AS_GEOHASHGRID_POLYGON) {
+      featureCollection.features.forEach((feature) => {
+        //replace geometries with the polygon
+        const esBbox = feature.properties.geohash_meta.rectangle;
```

Comment:
How about isolating this conversion functionality in https://github.com/elastic/kibana/blob/gis-plugin/x-pack/plugins/gis/public/elasticsearch_geo_utils.js and add some tests for it?